src/shared/fdset.c

```diff
@@ -97,21 +97,33 @@ int fdset_consume(FDSet *s, int fd) {
         return r;
 }
 
-int fdset_put_dup(FDSet *s, int fd) {
+int fdset_put_dup_full(FDSet *s, int fd, bool store_index) {
```

Comment:
frankly, i'd add a separate function for this.

int fdset_put_indexed() or so. i mean, i think flags parameters are generally OK, but I find them a bit icky if it changes completely what we return, i.e. an index rather than an fd.

Hence I'd just make these two functions. leave fdset_put_dup() as it is, and add a new function that does the index thing and only the index thing. the code of fdset_put_dup() is not so complicated, it's not that we'd duplicate huge pieces of code.